stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -16,28 +16,75 @@
 
 #include "swift/Runtime/Concurrency.h"
 #include "swift/ABI/Task.h"
+#include "swift/ABI/TaskGroup.h"
 #include "swift/ABI/Metadata.h"
+#include "swift/Runtime/Mutex.h"
 #include "swift/Runtime/HeapObject.h"
 #include "TaskPrivate.h"
+#include "TaskPrivate.h"
+#include "AsyncCall.h"
+#include "Debug.h"
+
+#include <dispatch/dispatch.h>
+
+#if !defined(_WIN32)
+#include <dlfcn.h>
+#endif
 
 using namespace swift;
-using GroupFragment = AsyncTask::GroupFragment;
+using TaskGroup = swift::TaskGroup;
 using FutureFragment = AsyncTask::FutureFragment;
 
-using ReadyQueueItem = GroupFragment::ReadyQueueItem;
-using ReadyStatus = GroupFragment::ReadyStatus;
-using GroupPollResult = GroupFragment::GroupPollResult;
+using ReadyQueueItem = TaskGroup::ReadyQueueItem;
+using ReadyStatus = TaskGroup::ReadyStatus;
+using PollResult = TaskGroup::PollResult;
+
+// =============================================================================
+// ==== create -----------------------------------------------------------------
+
+TaskGroup* swift::swift_task_group_create(AsyncTask *task) {
+  void *allocation = swift_task_alloc(task, sizeof(TaskGroup));
+
+  // nasty trick, but we want to keep the record inside the group as we'll need
+  // to remove it from the task as the group is destroyed, as well as interact
+  // with it every time we add child tasks; so it is useful to pre-create it here
+  // and store it in the group.
+  //
+  // The record won't be used by anyone until we're done constructing and setting
+  // up the group anyway.
+  void *recordAllocation = swift_task_alloc(task, sizeof(TaskGroupTaskStatusRecord));
+  auto record = new (recordAllocation)
+      TaskGroupTaskStatusRecord(reinterpret_cast<TaskGroup*>(allocation));
+
+  TaskGroup *group = new (allocation) TaskGroup(record);
+
+  // ok, now that the group actually is initialized: attach it to the task
+  swift_task_addStatusRecord(task, record);
+
+  return group;
+}
```

Comment:
I think this is fairly smart and safe, but would like to sanity check...

The group _and_ record are guaranteed to only exist as long as we're in the group, so we're ok to task allocate the record. The record we must store somewhere in order to later remove it -- removal is by `==`; as @rjmccall suggested; so we actually keep it in the group.

I would return a tuple of `(TaskGroup*, TaskGroupTaskStatusRecord*)` here but did not fiture out how to pass this nicely to swift... Would that be nicer than the trick of storing in the group? I don't mind the trick, open to opinions though... Would I have to make a struct to return here and declare the same shape in swift and that would be fine right?